vhost-user-blk implementation and minor fixes for vhost-user-backend

slp opened the pull request.

slp commented:
Reading #383, I've noticed we can do the same here. Please don't review this PR for the moment.

rbradford commented:
Thanks for the PR. :-)

@slp can you switch the integration tests (in `src/main.rs`) over to using your new backend.

Re: refactoring, it's your call what order you want to do it in, i'll happily approve (once the integration testing is enabled) if you want to follow-up with a refactor later.

slp commented:
@rbradford I've added the changes to switch to the new vhost-user-blk for the integration tests, and also refactored vhost_user_blk.rs to reuse as much code as possible. As a result, it has basically became a shim on top of vm-virtio, forcing us to add the missing advanced features to the latter, which I think it's the best possible scenario.

sameo commented on scripts/run_integration_tests.sh:
```diff
@@ -157,7 +155,7 @@ cp target/debug/cloud-hypervisor $VFIO_DIR
 newgrp kvm << EOF
 export RUST_BACKTRACE=1
-cargo test --features "integration_tests" -- --nocapture
+cargo test vhost_user_blk --features "integration_tests" -- --nocapture
```
You need to remove that change.

slp replied:
But this way is way faster! xD
Thanks for pointing to this, I've just removed the change.

sameo commented on scripts/run_integration_tests.sh:
```diff
@@ -168,7 +166,7 @@ if [ $RES -eq 0 ]; then
     newgrp kvm << EOF
 export RUST_BACKTRACE=1
-cargo test --features "integration_tests,mmio" -- --nocapture
+cargo test vhost_user_blk --features "integration_tests,mmio" -- --nocapture
```
Ditto.

sboeuf commented:
@rbradford I'll review this PR tomorrow, please don't merge it too quickly :)

sboeuf commented:
@slp could you please add a PR description as it is currently empty.

sboeuf commented on src/bin/vhost_user_blk.rs:
```diff
@@ -0,0 +1,299 @@
+        vrings: &[Arc<RwLock<Vring>>],
+    ) -> VhostUserBackendResult<bool> {
+        if evset != epoll::Events::EPOLLIN {
+            debug!("invalid events operation!\n");
```
Should be an `error!` or `warn!` instead of `debug!`

sboeuf commented on src/bin/vhost_user_blk.rs:
```diff
@@ -0,0 +1,299 @@
+
+        let mut vring = vrings[0].write().unwrap();
+        if self.process_queue(&mut vring) {
+            debug!("signaling queue");
```
s/signaling/signalling/ as the UK English is preferred.

sboeuf commented on src/bin/vhost_user_blk.rs:
```diff
@@ -0,0 +1,299 @@
+                    mem.write_obj(status, request.status_addr).unwrap();
+                }
+                Err(err) => {
+                    debug!("failed to parse available descriptor chain: {:?}", err);
```
Should be an `error!()`

sboeuf commented on src/bin/vhost_user_blk.rs:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    fn get_config(&self, _offset: u32, _size: u32) -> Vec<u8> {
+        let buf = unsafe {
```
It's better to add a comment why it's safe to invoke this `unsafe` block.

sboeuf commented on scripts/run_integration_tests.sh:
```diff
@@ -88,16 +88,14 @@ if [ ! -f "$VMLINUX_IMAGE" ]; then
 fi
 
 VIRTIOFSD="$WORKLOADS_DIR/virtiofsd"
```
Just a nit, but could you change the commit message title prefix `src:` into `ci:`?

sboeuf commented on vhost_user_backend/src/lib.rs:
```diff
@@ -477,9 +477,7 @@ impl<S: VhostUserBackend> VhostUserSlaveReqHandler for VhostUserHandler<S> {
     }
 
     fn set_features(&mut self, features: u64) -> VhostUserResult<()> {
-        if !self.owned {
```
Yes you're right!

sboeuf commented on src/main.rs:
```diff
@@ -1453,8 +1453,20 @@ mod tests {
+                    format!("image={},sock={}", image_path, sock_path).as_str(),
+                ])
+                .spawn()
+                .unwrap();
```
Instead of adding all this code, could you instead modify the `prepare_vubd()` function so that it spawns the `vhost-user-blk` daemon, using your new binary. This way the necessary changes should even be less intrusive.

sboeuf commented on src/main.rs:
```diff
@@ -1536,15 +1544,25 @@ mod tests {
-            );
+                ])
+                .spawn()
+                .unwrap();
```
Ditto. Please update `prepare_vubd()`.

sboeuf commented on src/bin/vhost_user_net.rs:
```diff
@@ -433,6 +433,10 @@ impl VhostUserBackend for VhostUserNetBackend {
             | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
     }
 
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
```
Yes that makes a lot of sense, I had forgot about that! Thanks for fixing it, we're gonna need it once we want to support complex features required by `vhost-user-fs`!

sboeuf commented:
@slp just wanted to know if you had time to work on the PR? If not, and because it's mostly ready, I could take care of the changes for you, just let me know.

slp commented:
@sboeuf Sorry, I'm travelling this week, so I planned to do the changes over the next one. Would it be too late?

sboeuf commented:
@slp 
>Sorry, I'm travelling this week, so I planned to do the changes over the next one. Would it be too late?

Oh no that's fine, no worries. I was just curious about ETA for this. But definitely next week will be fine ;)

slp commented:
@sboeuf PTAL, I think I addressed all issues.

sameo commented:
@slp Thanks. It seems the 2 vhost-user-blk tests are failing though. Are you able to run them locally?

rbradford commented:
@sboeuf Happy with the changes?

sboeuf commented:
@slp @rbradford I was in vacation last week. I'm happy to see we got this merged :)
@slp, thanks for your time rebasing on top `cloud-hypervisor` and `vhost_user_backend` crate!
